backend/api/src/redeem-market-ad-reward.ts

```diff
@@ -0,0 +1,74 @@
+import * as admin from 'firebase-admin'
+
+import { runRedeemBoostTxn } from 'shared/run-txn'
+import { createSupabaseDirectClient } from 'shared/supabase/init'
+import { z } from 'zod'
+import { APIError, authEndpoint, validate } from './helpers'
+
+const schema = z.object({
+  adId: z.string(),
+})
+
+const pg = createSupabaseDirectClient()
+
+export const redeemboost = authEndpoint(async (req, auth) => {
+  const { adId } = validate(schema, req.body)
+
+  pg.connect()
+  const firestore = admin.firestore()
+
+  // search txns if you've already reedemed this boost
+  const { count } = await pg.one(
+    `select count(*) from txns
+    where 
+    data->>'category' = 'MARKET_BOOST_REDEEM'
+    and data->>'fromId' = $1
+    and data->>'toId' = $2`,
+    [adId, auth.uid]
+  )
+
+  if (count > 1) {
+    throw new APIError(
+      403,
+      `You have already redeemed the boost for this market ${count} times`
+    )
+  }
+
+  // find the advertisement
+  const data = await pg.one(
+    `select cost_per_view::numeric, funds::numeric from market_ads
+    where id = $1`,
+    [adId]
+  )
+
+  const reward = parseFloat(data.cost_per_view)
+  const funds = parseFloat(data.funds)
```

Comment:
This is the one place where we should rely on the txns as source of truth and not just rely on the ad's stored funds.

If we fetched funds from txns within the redeem transaction below, that would actually make it correct.